accepted/future-releases/constant-update-2018/feature-specification.md

```diff
@@ -0,0 +1,158 @@
+When the operator is added, it should then also work in a constant expression.
+
+The `&`, `|` and `^` binary operators are now also allowed when the operands are of 
+type `bool`.
+
+### Short-Circuit Operators
+The `&&` operator is now short-circuit in constant and potentially constant expressions.
+It only attempts to evaluate the second operand if the first operand evaluates to `true`.
+This makes `false && (null as String).length` a valid constant expression.
+The second operand expression still needs to be a *potentially* constant expression,
+which is a new use of potentially constant expressions outside of const constructor
+initializer lists.
+
+Likewise the `||` operator only evaluates its second operand if the first evaluates to
+`false`, and the second operand must be a potentially constant expression.
+
+The `??` operator only evaluates its second operand if the first evaluates to
+`null`, and the second operand must be a potentially constant expression.
+
+Finally, the conditional `?`/`:` operator only evaluates one of its branches, 
+depending on whether the condition expression evaluates to `true` or `false`.
+The other branch must also be a potenatially constant expression.
+
+### Potentially Constant Expressions
+
+Potentially constant expressions were previously defined in terms of 
+"expressions that would be constant if constructor arguments are replaced with 
+constants of types matching the context", but something was a constant
+expression if it evaluated to a value without throwing. 
+
+That was not a functional definition, 
+so it has been changed to an entirely syntactic definition which allows 
+some syntactic constructs where the sub-expressions are also potentially constant. 
+There is no need to evaluate anything to determine if an expression is potentially constant,
+and there is no *type* requirements on the expression.
+
+The potentially constant expression must still satisfy all the same typing rules 
+as any other expression, independently of whether it's used at compile time.
+
+If a potentially constant expression is actually evaluated as part of a constant
+evaluation, then further type rules apply to the actual objects that are involved.
+
+*Notice*: That also means that there are obivous errors we do not catch.
+For example:
+```dart
+class C {
+  final x;
+  const C(List<int> x) : x = x.length;
+}
+```
+This class is accepted because `x.length` is a potentially constant expression
+independently of typing. However, `x.length` is only a constant expression if
+`x` is a `String`, so there is no possible valid constant invocation of this
+constructor.
+The constructor can still be invoked as a non-constant, though.
+
+### Clarifications
+
+The new specification clarifies that type variables cannot be used 
+as compile-time constant types, 
+and the type arguments to const constructor invocations and constant literals
+must be compile-time constant types.
+
+## Implementation Requirements
```

Comment:
This is expected to go into a second document named `implementation-plan.md`, see:
https://github.com/dart-lang/language/tree/master/accepted/2.1/super-mixins

Reply:
Will do.